src/main/java/com/artipie/rpm/ReactiveLock.java

```
@@ -0,0 +1,93 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.rpm;
+
+import io.reactivex.rxjava3.core.Completable;
+import io.reactivex.rxjava3.core.CompletableEmitter;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Reactive sync primitive.
+ *
+ * @since 0.1
+ */
+class ReactiveLock {
+
+    /**
+     * Unlock and emit sync object.
+     */
+    private final Object sync = new Object();
+
+    /**
+     * Current state of the lock.
+     */
+    private final AtomicBoolean locked = new AtomicBoolean(false);
+
+    /**
+     * Queue of lock acquires.
+     */
+    private final Queue<CompletableEmitter> acquires = new ConcurrentLinkedQueue<>();
+
+    /**
+     * Acquire the lock in a reactive way.
+     *
+     * @return Completion when lock is acquired.
+     */
+    public Completable lock() {
+        return Completable.create(
+            emitter -> {
+                this.acquires.add(emitter);
+                this.tryToEmitNext();
+            }
+        );
+    }
+
+    /**
+     * Unlock the lock.
+     */
+    public void unlock() {
+        synchronized (this.sync) {
+            if (this.locked.compareAndSet(true, false)) {
```

@Sammers21 I'd throw an exception in `else` branch - attempting to unlock non-locked state is a not valid case and signals about code level mistakes

---

Make sense.